add color to botany

joyalicegu opened the pull request:
Add ansi_render() to menu_screen.py to render ANSI-colored ASCII art.

Fall back on non-colored art if no colored art exists.

Only some escape codes for setting foreground color are supported.
Unsupported escape codes reset colors to normal.

Add ANSI-colored ASCII art for all plants to "art" directory.
Most of the art was colored using durdraw.
Add art/test.ansi for debugging purposes.

Also add a debugging mode for previewing art that can be enabled by
setting the "debug_art" flag to True in menu_screen.py. It adds a
"DEBUG" menu option that cycles through all art files.

(edited with updated commit message above)

---

Screenshots:

![screenshot-botany-dark](https://user-images.githubusercontent.com/16532904/147402677-d57fd9ed-e9e2-4e8b-90bd-bdedb3bf8488.png)
![screenshot-botany-light](https://user-images.githubusercontent.com/16532904/147402679-6fd4ca4e-bf38-4f42-9d63-00f8787bace3.png)
![screenshot-botany-jackolantern](https://user-images.githubusercontent.com/16532904/147402804-a38b79e8-9e43-43d9-a7d4-f788e1689814.png)
![screenshot-botany-rip](https://user-images.githubusercontent.com/16532904/147402805-65edec95-785a-4a7b-8f4d-0c443dc2f925.png)
![screenshot-botany-seed](https://user-images.githubusercontent.com/16532904/147402807-a63121bc-01ed-4ea7-8fbb-1d09fc45de14.png)
![screenshot-botany-seedling](https://user-images.githubusercontent.com/16532904/147402808-3a66918d-156c-4a21-b76f-f8587b198c18.png)

Demo videos of debug mode:

https://github.com/user-attachments/assets/923b8e95-1759-4e2b-bc06-b39525b07adf

https://github.com/user-attachments/assets/ddfccc17-4b01-40d0-988b-adb01b6bec75

joyalicegu commented:
This is really old, and I didn't finish coloring all of the existing ascii art. I will probably do some more work on it over the next couple of weeks (I should have an update to this PR ready by May).

I would like to test it more, too. I want to make sure it doesn't behave unexpectedly in different terminals / screen sizes / etc. Since this is just using the curses library, I don't expect anything weird to happen, but you never know. Maybe some other townies can try it out?

joyalicegu commented:
Sorry, it's been a while. I plan to continue this. Currently working on coloring the remaining art with durdraw.

Commit:
add color to botany

Add ansi_render() to menu_screen.py to render ANSI-colored ASCII art.

Fall back on non-colored art if no colored art exists.

Only some escape codes for setting foreground color are supported.
Unsupported escape codes reset colors to normal.

Add ANSI-colored ASCII art for all plants to "art" directory.
Most of the art was colored using durdraw.
Add art/test.ansi for debugging purposes.

Also add a debugging mode for previewing art that can be enabled by
setting the "debug_art" flag to True in menu_screen.py. It adds a
"DEBUG" menu option that cycles through all art files.

joyalicegu commented:
I've added all the remaining colored art for the plants. There's also a debug mode that allows cycling through art in botany for debugging purposes (set "self.debug_art = True" in menu_screen.py). I'll try to attach a demo video of the art in dark mode and light mode.

I did notice that plants have a "color" descriptor (in plant.py), and the color might not match that of the art displayed. Is that going to be a problem?

I think it would also be nice for people to be able to toggle this on and off (not sure where the setting would be saved though). Currently, the program just uses curses.has_colors().

joyalicegu commented:
Added demo videos of debug mode to the top comment, with all the colored art.

joyalicegu commented:
> I did notice that plants have a "color" descriptor (in plant.py), and the color might not match that of the art displayed. Is that going to be a problem?

Not sure what to do about this, thinking maybe I could have some designated color just be replaced by whatever color is in the color descriptor. e.g. green. Or the "color" could just sort of be a non-literal descriptor, and we could just ignore this. Or we could have a map for each plant where it shows where the color should be overridden by this color descriptor.

What do people think?

> I think it would also be nice for people to be able to toggle this on and off (not sure where the setting would be saved though). Currently, the program just uses curses.has_colors().

I guess this could be saved in the plant.json? And there could be an extra preferences menu, where this could be toggled. I don't want to make botany more complicated than it has to be though; a big part of its charm is its simplicity.

Does anyone have suggestions?

cc: @jifunks @vilmibm @noelleleigh

noelleleigh commented:
> > I think it would also be nice for people to be able to toggle this on and off (not sure where the setting would be saved though). Currently, the program just uses curses.has_colors().
> 
> I guess this could be saved in the plant.json? And there could be an extra preferences menu, where this could be toggled. I don't want to make botany more complicated than it has to be though; a big part of its charm is its simplicity.

A good start for allowing the user to control color output would be to respect [`NO_COLOR`](https://no-color.org/) and [`FORCE_COLOR`](https://force-color.org/) as [described in the Python documentation](https://docs.python.org/3/using/cmdline.html#controlling-color).

joyalicegu commented:
Thanks Noelle, I'll get to it soon!

On Thu, Jan 1, 2026 at 03:00 Noelle Leigh ***@***.***> wrote:

> *noelleleigh* left a comment (jifunks/botany#44)
> <https://github.com/jifunks/botany/pull/44#issuecomment-3703543795>
>
> I think it would also be nice for people to be able to toggle this on and
> off (not sure where the setting would be saved though). Currently, the
> program just uses curses.has_colors().
>
> I guess this could be saved in the plant.json? And there could be an extra
> preferences menu, where this could be toggled. I don't want to make botany
> more complicated than it has to be though; a big part of its charm is its
> simplicity.
>
> A good start for allowing the user to control color output would be to
> respect NO_COLOR <https://no-color.org/> and FORCE_COLOR
> <https://force-color.org/> as described in the Python documentation
> <https://docs.python.org/3/using/cmdline.html#controlling-color>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jifunks/botany/pull/44#issuecomment-3703543795>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD6ELKAAWYE55RTOSBDRFS34ET46PAVCNFSM6AAAAACMCHI4RCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBTGU2DGNZZGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

vilmibm commented:
Late to this but I continue to adore this work. Thank you so much.

Having a color pop that matches the description would be awesome but I wouldn't hold off from merging this if it's too onerous.

I do think supporting NO_COLOR/FORCE_COLOR is a great start.

joyalicegu committed:
respect color environment variables

joyalicegu commented:
OK, should respect color now! Also added a bit to the README about how to disable color (set `NO_COLOR` or `PYTHON_COLORS=0`). Added a TODO comment for if people want the color adjectives to be represented. (`"rainbow"` looks fun to handle haha)
